src/diff.jl

```diff
@@ -1,46 +1,30 @@
 # I do not use it but I import the function to add a method
 export differentiate
 
-differentiate(p::PolyVar, x)  = differentiate(Term(p), x)
-differentiate(p::Monomial, x) = differentiate(Term(p), x)
+# Fallback for everything else
+_diff_promote_op(::Type{T}, ::Type{<:AbstractVariable}) where T = T
+differentiate(α::T, v::AbstractVariable) where T = zero(T)
 
-function differentiate{C, T}(t::Term{C, T}, x::PolyVar{C})
-    i = findfirst(vars(t), x)
-    if i == 0 || t.x.z[i] == 0
-        S = Base.promote_op(*, T, Int)
-        zero(Term{C, S})
-    else
-        z = copy(t.x.z)
-        z[i] -= 1
-        Term(t.α * t.x.z[i], Monomial(vars(t), z))
-    end
-end
+_diff_promote_op(::Type{<:AbstractVariable}, ::Type{<:AbstractVariable}) = Int
+differentiate(v1::AbstractVariable, v2::AbstractVariable) = v1 == v2 ? 1 : 0
 
-function differentiate{C, T}(p::Polynomial{C, T}, x::PolyVar{C})
-    # grlex order preserved
-    i = findfirst(vars(p), x)
-    S = Base.promote_op(*, T, Int)
-    if i == 0
-        zero(Polynomial{C, S})
-    else
-        keep = find([z[i] > 0 for z in p.x.Z])
-        Z = [copy(p.x.Z[i]) for i in keep]
-        a = Vector{S}(length(keep))
-        for j in 1:length(Z)
-            a[j] = p.a[keep[j]] * Z[j][i]
-            Z[j][i] -= 1
-        end
-        Polynomial(a, MonomialVector(vars(p), Z))
-    end
-end
+_diff_promote_op(::Type{TT}, ::Type{<:AbstractVariable}) where {T, TT<:AbstractTermLike{T}} = changecoefficienttype(TT, Base.promote_op(*, T, Int))
+differentiate(t::AbstractTermLike, v::AbstractVariable) = coefficient(t) * differentiate(monomial(t), v)
+
+_diff_promote_op(::Type{PT}, ::Type{<:AbstractVariable}) where {T, PT<:APL{T}} = polynomialtype(PT, Base.promote_op(*, T, Int))
+differentiate(p::APL, v::AbstractVariable) = polynomial(differentiate.(terms(p), v))
```

Comment:
It is the responsability of `polynomial` to sort terms and remove zeros but currently `TypedPolynomials` does not remove zero terms, this should be fixed indeed :)